File: 7 trees/level order traversal.js
var levelOrder = function(root) {
    if(!root) return []
    const queue = [root] 
    const res = []
    while(queue.length !== 0){
       const queueLength = queue.length
       const level = []
       for(let i = 0; i < queueLength; i++){
           const current = queue.shift()
           if(current.left){
               queue.push(current.left)
           }
           if(current.right){
               queue.push(current.right)
           }
           level.push(current.val)
       }
       res.push(level)
   }
    return res
}

File: 7 trees/level order traversal.py
"""
O(n) time
O(n) space
"""
class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []
        q = []
    #initialize with root
        if root:
            q.append(root)
    #while queue is non-empty
        while q:
            level = []
            for i in range(len(q)):
                node = q.pop(0)
                level.append(node.val)
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            res.append(level)
        return res
  
#could queue instead of popleft()  
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        #result array
            res = []
        #queue is collections.deque
            q = collections.deque()
        #initialize with root
            if root:
                q.append(root)
        #while queue is non-empty
            while q:
                level = []
                for i in range(len(q)):
                    node = q.popleft()
                    level.append(node.val)
                    if node.left:
                        q.append(node.left)
                    if node.right:
                        q.append(node.right)
                res.append(level)
            return res
        
